src/io/csvReader.cu

```diff
@@ -0,0 +1,1230 @@
+
+	//-----------------------------------------------------------------------------
+	// free up space that is no longer needed
+	error = freeCSVSpace(raw_csv);
+	//checkError(error, "freeing raw_csv_t space");
+
+	//-----------------------------------------------------------------------------
+	// last step is to copy over the data - this also fills in gdf_data with the columns
+
+	//--- allocate space for the results
+	gdf_column **cols = (gdf_column **)malloc( sizeof(gdf_column *) * raw_csv->num_cols);
+
+
+	for (int col = 0; col < raw_csv->num_cols; col++) {
+
+		gdf_column *gdf = (gdf_column *)malloc(sizeof(gdf_column) * 1);
+
+		gdf->size		= raw_csv->num_records;
+		gdf->dtype		= raw_csv->dtypes[col];
+		gdf->null_count	= 0;						// will be filled in later
+
+		//--- column name
+		std::string str = raw_csv->col_names[col];
+		int len = str.length() + 1;
+		gdf->col_name = (char *)malloc(sizeof(char) * len);
+		memcpy(gdf->col_name, str.c_str(), len);
+		gdf->col_name[len -1] = '\0';
+
+		// allocate the CUDA memory
+		allocateGdfDataSpace(gdf);
+
+		launch_dataConvertColumn(raw_csv, info, gdf, col);
+		checkError(error, "call to launch_dataConvert " + col);
+
+		cols[col] = gdf;
+	}
+
+
+	//-----------------------------------------------------------------------------
+	//--- clean-up everything but the gdf_columns
+	error = freeInfoSpace(info);
+	checkError(error, "call to freeInfoSpace");
+	delete info;
+
+	error = freeCsvData(raw_csv->data);
+	checkError(error, "call to cudaFree(raw_csv->data)" );
+	delete raw_csv;
+
+	//printResults(cols, raw_csv->num_cols, 0, 10);
+
+	return cols;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------
+
+/**
+ * Since the column names remain on the host (CPU) side, they are saved as a std::string rather than char *
+ *
+ * This creates the csv_data_t structure which in turn creates the basic gdf_coulmn structure
+ *
+ */
+gdf_error parseColArguments(const std::vector<std::string>  &col_names, const std::unordered_map<std::string, std::string> &dtypes, const char delim, raw_csv_t *d)
+{
+
+	int num_col = col_names.size();						// how many columns are present
+
+	d->num_cols		= num_col;
+	d->num_records	= 0;
+	d->delimiter	= delim;
+
+	// start creating space for each column
+	for ( int x = 0; x < num_col; x++) {
+
+		std::string cn 	= col_names[x];
+		std::string dtA = dtypes.find(cn)->second;
+		gdf_dtype dt 	= convertStringToDtype( dtA );
+
+		//std::cout << cn << " == "  << dtA << "  ==  " << dt << std::endl;
+
+		if (dt == GDF_invalid || dt == GDF_STRING)
+			return GDF_UNSUPPORTED_DTYPE;
+
+		d->dtypes.push_back(dt);
+		d->col_names.push_back(cn);
+	}
+
+	return gdf_error::GDF_SUCCESS;
+}
+
+
+/*
+ * What is passed in is the data type as a string, need to convert that into gdf_dtype enum
+ */
+gdf_dtype convertStringToDtype(std::string &dtype) {
+
+	gdf_dtype dt = GDF_invalid;
+
+	if (dtype.compare( "str") == 0) {
```

Comment:
non-blocking but It would be better if this could be reformulated as a switch-case, rather than a bunch of if-else blocks.  I believe that dtypes get assigned a unique integer that you can use.